Sources/SwiftKafka/KafkaConsumer.swift

```
@@ -18,34 +18,71 @@ import NIOConcurrencyHelpers
+    }
+
+    func didTerminate() {
+        self.stateMachine.withLockedValue { $0.stopConsuming() }
+    }
 }
 
 // MARK: - KafkaConsumerMessages
 
 /// `AsyncSequence` implementation for handling messages received from the Kafka cluster (``KafkaConsumerMessage``).
 public struct KafkaConsumerMessages: Sendable, AsyncSequence {
+    let stateMachine: NIOLockedValueBox<KafkaConsumer.StateMachine>
+
     public typealias Element = KafkaConsumerMessage
     typealias BackPressureStrategy = NIOAsyncSequenceProducerBackPressureStrategies.NoBackPressure
-    typealias WrappedSequence = NIOAsyncSequenceProducer<Element, BackPressureStrategy, NoDelegate>
+    typealias WrappedSequence = NIOAsyncSequenceProducer<Element, BackPressureStrategy, KafkaConsumerCloseOnTerminate>
     let wrappedSequence: WrappedSequence
 
+    let logger: Logger
+
     /// `AsynceIteratorProtocol` implementation for handling messages received from the Kafka cluster (``KafkaConsumerMessage``).
     public struct ConsumerMessagesAsyncIterator: AsyncIteratorProtocol {
+        let stateMachine: NIOLockedValueBox<KafkaConsumer.StateMachine>
         var wrappedIterator: WrappedSequence.AsyncIterator
+        let logger: Logger
 
         public mutating func next() async -> Element? {
-            await self.wrappedIterator.next()
+            guard let element = await self.wrappedIterator.next() else {
+                return nil
+            }
+
+            let action = self.stateMachine.withLockedValue { $0.storeOffset() }
+            switch action {
+            case .storeOffset(let client):
+                do {
+                    try client.storeMessageOffset(element)
+                } catch {
+                    self.logger.error("""
+                    Failed committing offset '\(element.offset)' for partition '\(element.partition)'\
+                    of topic '\(element.topic)'
+                    """)
+                    return nil // Finishes sequence
```

IMO we should throw the error here instead of silently failing. Having this sequence be a throwing one is totally fine.